dashboard/lib/sample_data.rb

```diff
@@ -0,0 +1,181 @@
+include FactoryGirl::Syntax::Methods
+
+class SampleData
+  TEST_ACCOUNT_PASSWORD = "00secret".freeze
+  @@prng = nil
+
+  # Returns a seeded random number generator for consistent test data
+  def self.prng
+    @@prng ||= Random.new(0)
+    @@prng
+  end
+
+  # Creates sample data
+  def self.seed
+    raise "Should not be run outside of development" unless CDO.rack_env?(:development)
+
+    # Create a test teacher
+    teacher = create_teacher "testteacher@code.org", "TestTeacher Codeberg"
+
+    # Create normal-sized and large sections for each of CSF, CSD and CSP
+
+    create_section({teacher: teacher, name: "CSF 1",
+      login_type: Section::LOGIN_TYPE_PICTURE, grade: 2, age_min: 7,
+      age_max_inclusive: 9, script_name: "coursea-2018", num_students: 30,
+      use_imperfect_results: true}
+      )
+
+    create_section({teacher: teacher, name: "CSF 2 (Large)",
+      login_type: Section::LOGIN_TYPE_PICTURE, grade: 2, age_min: 7,
+      age_max_inclusive: 9, script_name: "coursea-2018", num_students: 150,
+      use_imperfect_results: true}
+      )
+
+    create_section({teacher: teacher, name: "CSD 1",
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 7, age_min: 13,
+      age_max_inclusive: 14, script_name: "csd1-2018", num_students: 30,
+      use_imperfect_results: false}
+      )
+
+    create_section ({teacher: teacher, name: "CSD 2 (Large)",
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 7, age_min: 13,
+      age_max_inclusive: 14, script_name: "csd1-2018", num_students: 150,
+      use_imperfect_results: false}
+      )
+
+    create_section ({teacher: teacher, name: "CSP 1",
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 10, age_min: 14,
+      age_max_inclusive: 16, script_name: "csp1-2018", num_students: 30,
+      use_imperfect_results: false}
+      )
+
+    create_section ({teacher: teacher, name: "CSP 2 (Large)",
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 10, age_min: 14,
+      age_max_inclusive: 16, script_name: "csp1-2018", num_students: 150,
+      use_imperfect_results: false}
+      )
+  end
+
+  # will delete the teacher and all of the teacher's sections and students
+  # and recreate.
+  def self.create_teacher(email, name)
+    # Delete any existing test data
+    user = User.find_by_email_or_hashed_email(email)
+    unless user.nil?
+      user.sections.each do |section|
+        # Delete all students in each section
+        section.followers.each do |follower|
+          student_user = User.find_by_id(follower.student_user_id)
```

Comment:
Unnecessary lookup. We can iterate [section.students](https://github.com/code-dot-org/code-dot-org/blob/a348a6ff89e66ec89910cd9db8d65cb1d3218559/dashboard/app/models/sections/section.rb#L56) directly since the `follower` here is only used to find the student. That will do one sql query with a join, rather than an initial query plus one each to find all the students.

Reply:
thx, done